design/embedded-cluster-cli-driven-update.md

```diff
@@ -0,0 +1,215 @@
+
+**User Experience Differences:**
+```bash
+# Non-semver channel: sequence-based ordering
+./binary update --license license.yaml
+→ "Update available: sequence 156 (was 155). Proceed? [y/N]"
+
+# Semver channel: semantic version ordering  
+./binary update --license license.yaml
+→ "Update available: v1.6.0 (from v1.5.2). Proceed? [y/N]"
+
+# Cross-channel semver comparison
+./binary update --version v1.7.0 --license license.yaml
+→ "Cannot update to v1.7.0 from beta channel. Required release v1.6.5 from stable channel must be installed first."
+```
+
+**Key Point:** This leverages existing infrastructure rather than building new version management from scratch.
+
+### 3. Update Discovery Mechanism
+
+The binary needs to discover available updates by calling the replicated.app API.
+
+**New API Endpoint:** `GET /embedded/:appSlug/:channelSlug/versions`
+- **Architecture**: Follow same proxy pattern as download endpoint (replicated-app → market-api)
+- **Authentication**: Same license-based authentication as download endpoint
+- **Authorization**: License ID in `Authorization` header
+- **Validation**: Same app/channel access controls and feature flag checks
+
+**API Response Format:**
+```json
+{
+  "channel": {
+    "id": "channel-123",
+    "name": "Stable", 
+    "slug": "stable",
+    "semverRequired": true,
+    "latestVersion": "1.2.3"
+  },
+  "versions": [
+    {
+      "versionLabel": "1.2.3",
+      "channelSequence": 456,
+      "ecVersion": "1.5.0+k8s-1.29",
+      "releaseDate": "2024-01-15T10:30:00Z",
+      "available": true,
+      "airgapSupported": true,
+      "required": false
+    },
+    {
+      "versionLabel": "1.2.2", 
+      "channelSequence": 455,
+      "ecVersion": "1.4.0+k8s-1.28",
+      "releaseDate": "2024-01-10T14:20:00Z",
+      "available": true,
+      "airgapSupported": true,
+      "required": false
+    }
+  ]
+}
+```
+
+**Key Implementation Requirements:**
+- **Reuse existing infrastructure**: Same authentication, validation, and proxy patterns as download endpoint
+- **Database optimization**: Efficient queries to avoid N+1 problems with large version lists
+- **Error handling**: Consistent error responses and HTTP status codes with download endpoint
+- **Performance**: Reasonable response times with appropriate caching headers
+
+### 4. Binary Download and Replacement (CLI-Driven Update Logic)
+
+**Current State:** No existing binary update logic in embedded cluster binary.
+
+#### Download Process
+Leverage existing embedded cluster download infrastructure:
+- **Reuse download endpoint**: Use current `/embedded/:appSlug/:channelSlug/:versionLabel` pattern
+- **Authentication**: Pass license ID from `--license` flag in Authorization header
+- **Streaming download**: Handle large binaries efficiently without excessive memory usage
+- **Progress indication**: Show download progress with percentage and transfer rates
```

Comment:
I think it is built on the fly so this may not be possible but worth keeping in here